### PR TITLE
feat(lambda): Migrate away from NODEJS_10_X and NODEJS_12_X to NODEJS_14_X

### LINKED ISSUE #19992: custom_resources: Node version 12 out of support
### Describe the bug

custom_resources builds a NodeJs lambda @ v12 of node which is out of support in 11 days.

### Expected Behavior

custom_resources should track LTS

### Current Behavior

tracks maintenance/deprecated version

### Reproduction Steps

use module

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

all

### Framework Version

all

### Node.js Version

12

### OS

all

### Language

Typescript

### Language Version

3.8.3

### Other information

_No response_

### PR DESCRIPTION
**Pull request by robertd**
Spiritual successor of #20531

Screen Shot 2022-06-01 at 9 08 18 PM

- Migrated away (tests, docs, internal use) from NODEJS_10_X to NODEJS_14_X
- Migrated away (tests, docs, internal use) from NODEJS_12_X to NODEJS_14_X
- Added NODEJS_16_X to CustomResources
- Update CustomResources to use nodejs14.x by default
- Updated integration tests across the board. Some integ tests introduced destructive changes (expected with change of nodejs runtime)

Closes #20568
Closes #19992
Closes #20474

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by robertd** [comment]
- Migrated away (tests, docs, internal use) from NODEJS_10_X to NODEJS_14_X
- Migrated away (tests, docs, internal use) from NODEJS_12_X to NODEJS_14_X
- Added NODEJS_16_X to CustomResources
- Updated integration tests across the board. Some integ tests introduced destructive changes (expected with change of nodejs runtime)

**Comment by robertd** [comment]
@jogold @pahud @rix0rrr Can you please review and fast-track this PR since it touches so many modules? Thanks!

File: tools/@aws-cdk/prlint/action.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 name: Pull Request Linter
 description: Execute validation rules on GitHub Pull Requests
 runs:
-  using: node12
+  using: node14
```
Comment (pahud) [inline tools/@aws-cdk/prlint/action.yml:4]: Looks like `node14` is not supported here?

Comment (robertd) [reply tools/@aws-cdk/prlint/action.yml:4]: Yep, you are correct. I'll revert this really quick. https://github.com/actions/runner/issues/772

**Comment by pahud** [comment]
I didn't see any specific issues. The core team should look into this PR very soon. Thank you for yet another awesome PR @robertd !

File: packages/aws-cdk/does-not-exist.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"expiration":1654210101399,"notices":[{"title":"Error when building EKS cluster with monocdk import","issueNumber":17061,"overview":"When using monocdk/aws-eks to build a stack containing an EKS cluster, error is thrown about missing lambda-layer-node-proxy-agent/layer/package.json.","components":[{"name":"cli","version":"<1.130.0 >=1.126.0"}],"schemaVersion":"1"}]}
```
Comment (kaizencc) [inline packages/aws-cdk/does-not-exist.json:1]: Lets remove this file

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/eventbridge/put-events.integ.snapshot/asset.41fc8f2dc7c01b34dda9916c7f763e7b7909eb629da9ffe879cb786114aae736.bundle/index.js [hunk 3]
```diff
@@ -509,27 +512,9 @@ var AssertionHandler = class extends CustomResourceHandler {
...
-        this.type = "stringLikeRegexp";
-        this.parsedObj = obj.$StringLike;
-        break;
-      default:
-        this.type = "exact";
-        this.parsedObj = obj;
-    }
+    this.parsedObj = {
+      matcher: obj
+    };
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/eventbridge/put-events.integ.snapshot/asset.41fc8f2dc7c01b34dda9916c7f763e7b7909eb629da9ffe879cb786114aae736.bundle/index.js:517]: The changes in this file don't look related to the nodejs upgrades, and I'm too lazy to figure out what this is. Can you explain what this is / why its necessary?

File: packages/@aws-cdk/aws-lambda/test/runtime.inlinecode.integ.snapshot/aws-cdk-lambda-runtime-inlinecode.template.json [hunk 4]
```diff
@@ -31,26 +31,26 @@
     ]
    }
   },
-  "NODEJS12X8B8075A4": {
+  "PYTHON364935EF15": {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda/test/runtime.inlinecode.integ.snapshot/aws-cdk-lambda-runtime-inlinecode.template.json:34]: why did this one change to python?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json [hunk 5]
```diff
@@ -14,28 +14,28 @@
         }
       }
     },
-    "3a0df356c74318460e8172175150c9d21465d77dcfdda6f7686b4b10af25db19": {
+    "45851fd5f84b5067f2232644bbf86f53d313030a3fca7b2e0282f2017629dfc1": {
       "source": {
-        "path": "/home/hallcor/work/aws-cdk-fork",
+        "path": "/Users/rdjurasaj/code/robertd/aws-cdk",
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json:19]: dang, what is going on here! might not be anything you can change but seems weird.

File: packages/@aws-cdk-containers/ecs-service-extensions/test/assign-public-ip.integ.snapshot/tree.json [hunk 6]
```diff
@@ -642,120 +642,13 @@
                   }
                 },
                 "constructInfo": {
-                  "fqn": "@aws-cdk/core.CfnResource",
-                  "version": "0.0.0"
-                }
-              },
-              "ExecutionRole": {
```
Comment (kaizencc) [inline packages/@aws-cdk-containers/ecs-service-extensions/test/assign-public-ip.integ.snapshot/tree.json:649]: why was all this deleted?

**Review by kaizencc — APPROVED** [review]
Hi @robertd, thanks for doing this! Looks like the changes fall into the following categories:

- integ test version updates -> these are all fine
- hash changes -> also fine
- updating lambda runtimes -> the manual work you did, also great
- weird things here and there that I tried to call out. not necessarily anything you can do, but wanted to call attention to them.
- the `does-not-exist.json` file, which shouldn't be in this PR :).

Want to get this in fast, so I'm going to investigate a few of my own comments, hope that's ok.

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/eventbridge/put-events.integ.snapshot/asset.41fc8f2dc7c01b34dda9916c7f763e7b7909eb629da9ffe879cb786114aae736.bundle/index.js (hunk 3 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/eventbridge/put-events.integ.snapshot/asset.41fc8f2dc7c01b34dda9916c7f763e7b7909eb629da9ffe879cb786114aae736.bundle/index.js:517]: Well looks like all of these changes come from the `integ-tests` module which has had lots of changes lately, so i'm not going to pay it any mind.

File: packages/@aws-cdk/aws-lambda/test/runtime.inlinecode.integ.snapshot/aws-cdk-lambda-runtime-inlinecode.template.json (hunk 4 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-lambda/test/runtime.inlinecode.integ.snapshot/aws-cdk-lambda-runtime-inlinecode.template.json:34]: it didn't. these are just poorly rendered diffs

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json (hunk 5 above)
Comment (jogold) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json:19]: I think this is because asset staging is disabled when running integ tests and this particular tests uses a `NodejsFunction`. 

With bundling when asset staging is disabled the stage path is the source path and for a `NodejsFunction` the source path is the path where the lock file resides (so docker can access all node modules).

Comment (jogold) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json:19]: https://github.com/aws/aws-cdk/blob/49b397c343b1a2581b2f4d2c098729b21b4b0290/packages/%40aws-cdk/core/lib/asset-staging.ts#L97

Comment (kaizencc) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/test/s3-bucket.lit.integ.snapshot/aws-cdk-firehose-delivery-stream-s3-all-properties.assets.json:19]: perfect, thanks

File: packages/@aws-cdk-containers/ecs-service-extensions/test/assign-public-ip.integ.snapshot/tree.json (hunk 6 above)
Comment (kaizencc) [reply packages/@aws-cdk-containers/ecs-service-extensions/test/assign-public-ip.integ.snapshot/tree.json:649]: guess the tree has changed since this was last updated. probably not worth really diving into more than that.

**Comment by robertd** [comment]
@kaizencc & @jogold Thanks for stepping in and taking a look. Much appreciated. New `integ` was really messing with me at the end there since I've been doing this bulky PR three times now :D. I believe new `integ` workflows require some more polishing down the road (my personal observation), as I was getting some changes introduced that I wasn't able to explain logically where they are coming from. New `integ` process is definitely more verbose and touches larger file surface than the process before. I was running ` yarn integ-runner --update-on-failed --dry-run --parallel-regions us-west-2` in every module that was affected by this change.

Again... thanks for stepping in and pushing this across the finish line. ❤️

**Comment by McDoit** [comment]
Would be utterly amazing to get this backported to CDKv1, as migrations to v2 might take time but deploys will stop working for older nodejs versions on v1...

**Comment by tmitchel2** [comment]
This is causing failures because some aws code build images dont support nodejs 14 yet as per https://github.com/aws/aws-cdk/issues/20739

**Comment by NickDarvey** [comment]
Do you know when this will be release @robertd? I can't see this PR referenced in [CHANGELOG](https://raw.githubusercontent.com/aws/aws-cdk/main/CHANGELOG.md) and AWS is hassling me about out-of-support Lambda functions...

**Comment by jacobtolar** [comment]
@NickDarvey Looks like it is in 2.28.0 or higher ( [CHANGELOG.v2.md](https://github.com/aws/aws-cdk/blob/main/CHANGELOG.v2.md#2280-2022-06-14) ) but was never backported to 1.x

**Comment by robertd** [comment]
What @jacobtolar said. I'd suggest updating to CDK v2 if you can.
